README.md

```diff
@@ -204,6 +204,84 @@ The following pieces of container metadata are available to plugins in NRI:
 Apart from data identifying the container, these pieces of information
 represent the corresponding data in the container's OCI Spec.
 
+## Pod Network Lifecycle Events
+
+NRI plugins can subscribe to the following pod network events:
+
+  - Pod network configuration change
+  - Pod network pre-creation before CNI plugins are called
+  - Pod network post-creation after CNI plugins have been called
+  - Pod network removal
+
+### Pod Network Configuration Change
+
+<details>
+<summary>NRI pod network configuration change event</summary>
+<p align="center">
+<img src="./docs/nri-cni-conf.svg" title="NRI Pod Network Configuration Change Event">
+</p>
+</details>
+
+A NetworkConfigurationChanged event is sent each time a change in CNI
+configuration is detected. It is left as an implementation detail for the
+runtime whether a NetworkConfigurationChanged evet contains one or more
+changed CNI configurations should a simultaneous modification of more than
+one configuration be detected at the same time.
+
+NRI plugins are able to alter only those CNI configuraions that are sent by the
+runtime, no additions are allowed. An NRI plugin can return either a modified
+version of one or more CNI configurations received, unmodified versions or an
+empty value or nil. Returned values will later be used by the runtime when a
+pod network is created.
+
+### Pod Network Creation
+
+<details>
+<summary>NRI pod network setup events</summary>
+<p align="center">
+<img src="./docs/nri-cni-setup.svg" title="NRI Pod Network Setup Events">
+</p>
+</details>
+
+### PreSetupNetwork
+
+The PreSetupNetwork event is sent immediately before CNI is invoked to set
+up the pod network in its namespace. In addition to the PodSandbox object,
+arguments to PreSetupNetwork includes an array of strings containing all JSON
+formatted CNI configuration files associated with the pod.
+
+The reply data consists of CNI capabilities and 'args' arguments. Any returned
```

Comment:
"CNI capabilities" is not a grea API, mainly because is a convention https://github.com/containernetworking/cni/blob/main/CONVENTIONS.md#current-conventions

Reply:
Looking at the above comments, we need plain 'opts' transported back and forth here?